edb/server/protocol/auth_ext/_static/webauthn-authenticate.js

```diff
@@ -0,0 +1,188 @@
+    return;
+  }
+
+  // Fetch WebAuthn options from the server
+  const options = await getAuthenticateOptions(email);
+
+  // Get the existing credentials assertion
+  const assertion = await navigator.credentials.get({
+    publicKey: {
+      ...options,
+      challenge: decodeBase64Url(options.challenge),
+      allowCredentials: options.allowCredentials.map((credential) => ({
+        ...credential,
+        id: decodeBase64Url(credential.id),
+      })),
+    },
+  });
+
+  // Register the credentials on the server
+  const registerResult = await authenticateAssertion({
+    email,
+    assertion,
+    challenge,
+  });
+
+  return registerResult.code ?? null;
+}
+
+/**
+ * Fetch WebAuthn options from the server
+ * @param {string} email - Email address to register
+ * @returns {Promise<globalThis.PublicKeyCredentialCreationOptions>}
+ */
+async function getAuthenticateOptions(email) {
+  const url = new URL(WEBAUTHN_OPTIONS_URL);
+  url.searchParams.set("email", email);
+
+  const optionsResponse = await fetch(url, {
+    method: "GET",
+  });
+
+  if (!optionsResponse.ok) {
+    console.error(
+      "Failed to fetch WebAuthn options:",
+      optionsResponse.statusText
+    );
+    console.error(await optionsResponse.text());
+    throw new Error("Failed to fetch WebAuthn options");
+  }
+
+  try {
+    return await optionsResponse.json();
+  } catch (e) {
+    console.error("Failed to parse WebAuthn options:", e);
+    throw new Error("Failed to parse WebAuthn options");
+  }
+}
+
+/**
+ * Authenticate the credentials on the server
+ * @param {Object} props
+ * @param {string} props.email
+ * @param {Object} props.assertion
+ * @param {string} props.provider
+ * @param {string} props.challenge
+ * @returns {Promise<Object>}
+ */
+async function authenticateAssertion(props) {
+  // Assertion includes raw bytes, so need to be encoded as base64url
+  // for transmission
+  const encodedAssertion = {
+    ...props.assertion,
+    rawId: encodeBase64Url(new Uint8Array(props.assertion.rawId)),
+    response: {
+      ...props.assertion.response,
+      authenticatorData: encodeBase64Url(
+        new Uint8Array(props.assertion.response.authenticatorData)
+      ),
+      clientDataJSON: encodeBase64Url(
+        new Uint8Array(props.assertion.response.clientDataJSON)
+      ),
+      signature: encodeBase64Url(
+        new Uint8Array(props.assertion.response.signature)
+      ),
+      userHandle: props.assertion.response.userHandle ? encodeBase64Url(
+        new Uint8Array(props.assertion.response.userHandle)
+      ) : null,
+    },
+  };
+
+  const authenticateResponse = await fetch(WEBAUTHN_AUTHENTICATE_URL, {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify({
+      email: props.email,
+      assertion: encodedAssertion,
+      provider: props.provider,
+      challenge: props.challenge,
+    }),
+  });
+
+  if (!authenticateResponse.ok) {
+    console.error(
```

Comment:
I don't think this handles the case where email verification is required, so we'll end up returning a misleading error that webauthn auth failed.

Reply:
Good point. Now that we have stable type names in our error json, I can detect this case and at least throw a more helpful error. We still navigate you to the "error" state of the sign in page in the email+password case, so I think that throwing the correct error should do the right thing.

Reply:
@jaclarke ⚡ [`64eaa1e` (#6792)](https://github.com/edgedb/edgedb/pull/6792/commits/64eaa1ec349562a66d7e38a65f988bbc3ffd3075)